File: Week 5/python/3.recursion_basics.py
"""
WEEK 5 - PYTHON FUNCTIONS & RECURSION
Topic: Recursion Basics
File: 3.recursion_basics.py

CONCEPT:
A recursive function calls itself with a SMALLER instance of the problem.
Every recursion needs:
  1. BASE CASE -- terminates the recursion
  2. RECURSIVE CASE -- progresses toward the base case

KEY POINTS:
 - Python's default RECURSION LIMIT is 1000 (sys.setrecursionlimit to change).
 - No tail-call optimisation -- deep recursion -> RecursionError.
 - For deep recursion, prefer iteration or convert to a stack-based loop.
 - functools.lru_cache memoises results, turning O(2^n) -> O(n).

PROBLEMS COVERED:
 1. Print N down to 1
 2. Print 1 up to N (using post-order)
 3. Factorial
 4. Sum of first N
 5. Power (linear and fast exponentiation)

COMPLEXITY:
 - Linear recursions: O(n) time, O(n) call-stack space
 - Fast power: O(log n) time
"""

import sys


def print_desc(n: int) -> None:
    if n == 0:
        return
    print(n, end=' ')
    print_desc(n - 1)


def print_asc(n: int) -> None:
    if n == 0:
        return
    print_asc(n - 1)
    print(n, end=' ')


def factorial(n: int) -> int:
    if n <= 1:
        return 1
    return n * factorial(n - 1)


def sum_n(n: int) -> int:
    if n == 0:
        return 0
    return n + sum_n(n - 1)


def power(base: int, exp: int) -> int:
    if exp == 0:
        return 1
    return base * power(base, exp - 1)


def fast_power(base: int, exp: int) -> int:
    if exp == 0:
        return 1
    if exp % 2 == 0:
        half = fast_power(base, exp // 2)
        return half * half
    return base * fast_power(base, exp - 1)


def main() -> None:
    print("Descending 5..1: ", end='')
    print_desc(5)
    print()

    print("Ascending 1..5: ", end='')
    print_asc(5)
    print()

    print(f"5!  = {factorial(5)}")
    print(f"10! = {factorial(10)}")

    print(f"sum(10) = {sum_n(10)}")

    print(f"2^10 = {power(2, 10)}")
    print(f"2^10 (fast) = {fast_power(2, 10)}")
    print(f"3^20 (fast) = {fast_power(3, 20)}")

    # Demonstrate recursion limit
    sys.setrecursionlimit(2000)
    print(f"sum_n(1500) = {sum_n(1500)}")


if __name__ == "__main__":
    main()


# NOTES:
# - Java/C++ have larger default stack sizes; Python's is only ~1000 frames.
# - For sum/factorial use the closed-form: n*(n+1)//2 / math.factorial(n).
# - Fast power uses exponent-by-squaring -- a fundamental DSA pattern.
# - Convert recursion to a loop or an explicit stack to avoid RecursionError.

File: Week 5/cpp/3.recursion_basics.cpp
/*
 * WEEK 5 - C++ FUNCTIONS & RECURSION
 * Topic: Recursion Basics
 * File: 3.recursion_basics.cpp
 *
 * CONCEPT:
 *  Recursive functions invoke themselves with a smaller problem instance.
 *  Need a BASE CASE and a RECURSIVE CASE.
 *
 * KEY POINTS:
 *  - C++ has no built-in tail-call optimisation guarantee (some compilers do it).
 *  - Stack typically has 1-8 MB -- recursion depths into the millions can crash.
 *  - Prefer iteration for very deep problems.
 *
 * COMPLEXITY:
 *  - Linear recursion: O(n) time, O(n) stack.
 *  - Fast power: O(log n).
 */

#include <iostream>

void print_desc(int n) {
    if (n == 0) return;
    std::cout << n << ' ';
    print_desc(n - 1);
}

void print_asc(int n) {
    if (n == 0) return;
    print_asc(n - 1);
    std::cout << n << ' ';
}

long long factorial(int n) {
    if (n <= 1) return 1;
    return (long long)n * factorial(n - 1);
}

int sum_n(int n) {
    if (n == 0) return 0;
    return n + sum_n(n - 1);
}

long long power(long long base, int exp) {
    if (exp == 0) return 1;
    return base * power(base, exp - 1);
}

long long fast_power(long long base, int exp) {
    if (exp == 0) return 1;
    if (exp % 2 == 0) {
        long long h = fast_power(base, exp / 2);
        return h * h;
    }
    return base * fast_power(base, exp - 1);
}

int main() {
    std::cout << "Descending 5..1: "; print_desc(5); std::cout << '\n';
    std::cout << "Ascending  1..5: "; print_asc(5);  std::cout << '\n';

    std::cout << "5!  = " << factorial(5) << '\n';
    std::cout << "10! = " << factorial(10) << '\n';

    std::cout << "sum(10) = " << sum_n(10) << '\n';

    std::cout << "2^10 = " << power(2, 10) << '\n';
    std::cout << "2^10 (fast) = " << fast_power(2, 10) << '\n';
    std::cout << "3^20 (fast) = " << fast_power(3, 20) << '\n';

    return 0;
}

/*
 * NOTES:
 *  - long long is 64-bit; factorial(20) fits, factorial(21) overflows.
 *  - For very large factorials, use Boost.Multiprecision.
 *  - Fast exponentiation by squaring is a foundational DSA technique.
 */
